### ISSUE TITLE
LRU .values() and dict return old entries

### ISSUE THREAD

**Issue by nicoddemus** [mahmoud/boltons#348]
Hi,

First of all thanks for the excellent library!

I'm seeing strange results when using `LRU`: after replacing an existing entry with a new value, seems like the `keys()` and `values()` methods return the old value, instead of the new one. Using `__getitem__` and `.get` we have the expected behavior.

Here is a test showing this behavior:

```python
def test_lru():
    from boltons.cacheutils import LRU

    cache = LRU()

    # Add an entry.
    cache['a'] = 1
    
    # Normal __getitem__ access. 
    assert cache['a'] == 1  # passes.
    # Convert to dict.
    assert dict(cache) == {'a': 1}  # passes.
    # Another way to access the only value.
    assert list(cache.values())[0] == 1  # passes.

    # Replace the existing 'a' entry with a new value.
    cache['a'] = 200

    # __getitem__ works as expected.
    assert cache['a'] == 200  # passes.

    # Both dict and accessing via values() return the old entry: 1.
    assert dict(cache) == {'a': 200}  # fails.
    assert list(dict(cache).values())[0] == 200  # fails.
```

This test fails in the last asserts (of course to see the 2nd failure one needs to comment the first):

```pytest
λ pytest foo.py
======================== test session starts ========================
platform win32 -- Python 3.10.9, pytest-7.4.2, pluggy-1.3.0
rootdir: c:\Users\bruno\projects\boltons
configfile: pytest.ini
collected 1 item

foo.py F                                                       [100%]

============================= FAILURES ==============================
_____________________________ test_lru ______________________________

    def test_lru():
        from boltons.cacheutils import LRU

        cache = LRU()

        # Add an entry.
        cache['a'] = 1

        # Normal __getitem__ access.
        assert cache['a'] == 1  # passes.
        # Convert to dict.
        assert dict(cache) == {'a': 1}  # passes.
        # Another way to access the only value.
        assert list(cache.values())[0] == 1  # passes.

        # Replace the existing 'a' entry with a new value.
        cache['a'] = 200

        # __getitem__ works as expected.
        assert cache['a'] == 200  # passes.

        # Both dict and accessing via values() return the old entry: 1.
>       assert dict(cache) == {'a': 200}  # fails.
E       AssertionError: assert {'a': 1} == {'a': 200}
E         Differing items:
E         {'a': 1} != {'a': 200}
E         Use -v to get more diff

foo.py:23: AssertionError
====================== short test summary info ======================
FAILED foo.py::test_lru - AssertionError: assert {'a': 1} == {'a': 200}
========================= 1 failed in 0.04s =========================
```

At first it seems like a bug? Or perhaps this usage is not meant to be supported?

**Comment by mahmoud** [comment]
Hey Bruno! Good to see you here again. Thanks for the kind words and detailed report. I confirmed I'm seeing the same behavior. 

I never really had that usage pattern IRL, but I think it's a great one to support. I've got a fix on a branch that makes it work. I just pushed to #349. Still need to think about it a bit, but feel free to take a look!

**Comment by mahmoud** [comment]
Fix is now up in 23.1.0. Thanks again for the report/review!

**Comment by nicoddemus** [comment]
Awesome, thanks!
